### ISSUE TITLE
Poor (and nearly-incorrect) example for rule F.27 (shared_ptr)

### ISSUE THREAD

**Issue by eyalroz** [isocpp/CppCoreGuidelines#2171]
At [F.27: Use a shared_ptr<T> to share ownership](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-shared_ptr) , we have the following example:
```
shared_ptr<const Image> im { read_image(somewhere) };

std::thread t0 {shade, args0, top_left, im};
std::thread t1 {shade, args1, top_right, im};
std::thread t2 {shade, args2, bottom_left, im};
std::thread t3 {shade, args3, bottom_right, im};

// detach threads
// last thread to finish deletes the image
```
It's not true that the last thread from among t0...t3 deletes the image; the code which creates them still has a shared_ptr. Now, it's true that this code is also part of a thread - but the surrounding text does not stress this point, and the comment gives the impression that it refers to threads t0...t3. In fact, it is possible that the creating thread has other work, and doesn't need to / want to keep the image alive.

Which brings me to another point: If the creating thread _does_ intend to do more work on the image later on, then

* If the work depends on the work of t0...t3, then a shared_ptr doesn't make sense to begin with: It needs to be a vanilla automatic-storage `Image` object, then a joining on the 4 threads, then more work on the image. That's the use of the suggested Alternative in F.27.
* If the work doesn't depend on that of t0...t3, it's not clear why the creating thread is doing that work, rather than detaching a thread t4 to do it.
* If the work doesn't depend on that of t0...t3, then this should be indicated clearly in the example code, e.g. by invoking some kind of `do_work(im)` function after detaching the threads.

PS - Nitpick: `im` is a poor choice of name; I would change it to `image`. That would be in line with other variable names in the example, like `shade`.

**Comment by hsutter** [comment]
Editors call: Thanks! We agree this would be clearer with a destruction of `im`.

**Comment by eyalroz** [comment]
> We agree this would be clearer with a destruction of im.

Something like this perhaps?
```
{
  shared_ptr<const Image> im { read_image(somewhere) };

  std::thread t0 {shade, args0, top_left, im};
  std::thread t1 {shade, args1, top_right, im};
  std::thread t2 {shade, args2, bottom_left, im};
  std::thread t3 {shade, args3, bottom_right, im};

  // detach threads
}

// main thread no longer shares ownership, last among t0-t3 to finish deletes the image
```
or maybe even
```
{
  shared_ptr<const Image> im { read_image(somewhere) };

  std::thread{shade, args0, top_left, im}.detach();
  std::thread{shade, args1, top_right, im}.detach();
  std::thread{shade, args2, bottom_left, im}.detach();
  std::thread{shade, args3, bottom_right, im}.detach();
}
// main thread no longer shares ownership, last among t0-t3 to finish deletes the image
```
?

**Comment by hsutter** [comment]
Yes, that's pretty much the direction we had in mind 👍

**Comment by eyalroz** [comment]
@hsutter : So, this bug is not actually resolved. After the new commit, the code as written results in `std::terminate` being called -  as there is now not even a placeholder for the threads getting detached. You're saying that "detaching thread requires care etc." - but the detaching is not said to occur. 

This will confuse users (and doubly so if they mistakenly imagine that the destruction involves joining - which they might believe happens; under that mis-perception, the use-case would actually be better served by a plain scoped `Image` rather than the shared_ptr. It's better to be extra clear in distinguishing the intended use case from the very-typical false use case for smart pointers.

Bottom line: Please either actually detach (which isn't hard after all), or say the threads get detached, explicitly.
